### DISCUSSION TITLE
Copy object from one relationship to another

### DISCUSSION THREAD

**Discussion by santalvarez** [sqlalchemy/sqlalchemy#9453] [Usage Questions] [answered]
Hi I am experiencing a really weird behavior and I can't figure out why.
I have this two simple objects:
```python
class Group(db.Model):
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    name = Column(String, nullable=False)

    users = relationship("User", back_populates="group")

    def update(self, new: Group) -> None:
        self.users = []
        new.users[0].group_id = self.id
        self.users.append(new.users[0])

class User(db.Model):
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    group_id = Column(BigInteger, ForeignKey("group.id"), nullable=False)
    name = Column(String, nullable=False)

    group = relationship("Group", back_populates="users")
```
And I want to test the `update` function of `Group`. I am using this simple test for that:
```python
def test_update_group():
    old = Group(id=1,
                name="group1",
                users=[User(id=1, group_id=1, name="user1")]
    )
    new = Group(id=2,
                name="group2",
                users=[User(id=2, group_id=2, name="user2")]
    )

    old.update(new)

    # Here new.users is [] why?????????
    assert old.users[0].id == new.users[0].id

```
The problem is after i append the user of `new` to `old` the users of `new` become empty. Why is this happening? 

Thanks in advance.

**Comment by zzzeek** [comment, marked as answer]
you have back populates set up, so this:

    self.users.append(new.users[0])

expands to this:

    u = new.users[0]   # get existing User(id=2)
    self.users.append(u)   # append to "old.users"
    u.group = self  # backref sets User(id=2).group to "old"
    new.users.remove(u)  # previous group named "new" is de-associated from User(id=2)
   
more succinctly it can be seen as:

    new.users[0].group = old
    assert not new.users

**Reply by santalvarez** [reply]
Ohh I see. Thanks a lot!
